include/fi_perf.h

```diff
@@ -0,0 +1,386 @@
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_PERF_H_
+#define _FI_PERF_H_
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <inttypes.h>
+#include <stdio.h>
+
+/* Generic logic */
+
+#define FI_PERF_SLOT_COUNT_MAX       (8)
+#define FI_PERF_SLOT_NAME_LENGTH_MAX (256)
+#define FI_PERF_PRIVATE_AREA_SIZE    (64)
+
+struct fi_perf_slot {
+	uint64_t begin;
+	uint64_t sum;
+	uint64_t count;
+	uint64_t is_active;
+};
+
+struct fi_perf_handle {
+	struct fi_perf_slot slots[FI_PERF_SLOT_COUNT_MAX];
+	char slot_names[FI_PERF_SLOT_COUNT_MAX][FI_PERF_SLOT_NAME_LENGTH_MAX];
+	int count;
+	uint8_t private[FI_PERF_PRIVATE_AREA_SIZE];
+};
+
+#ifdef FI_PERF_RDPMC
+
+/*
+ * The following is a modification/adaptation of a code from Andi Kleen
+ * pmu-tools project:
+ *  https://github.com/andikleen/pmu-tools/blob/master/jevents/rdpmc.c
+*/
+
+/*
+ * Copyright (c) 2012,2013 Intel Corporation
+ * Author: Andi Kleen
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that: (1) source code distributions
+ * retain the above copyright notice and this paragraph in its entirety, (2)
+ * distributions including binary code include the above copyright notice and
+ * this paragraph in its entirety in the documentation or other materials
+ * provided with the distribution
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+/* RDPMC specific logic */
+
+#include <unistd.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/fcntl.h>
+#include <linux/perf_event.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <asm/unistd.h>
+
+struct fi_perf_rdpmc_context {
+	int fd;
+	struct perf_event_mmap_page *buf;
+	uint32_t type;
+	uint32_t config;
+};
+
+#if defined(__ICC) || defined(__INTEL_COMPILER)
+#include "immintrin.h"
+#endif
```

Comment:
nit: can we move this include up before the structure definition?

Reply:
I include that in order to get ```_rdpmc``` with icc. I've moved it down closer to the place where I need that but I may move it to the top so it would be grouped with other headers